fix: pass all valid options to nested commands in `ddev composer create`, fixes #6300, fixes #6246, for #5058

stasadev opened the pull request:
<!-- 
  PR titles have very precise rules, please read 
  https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-pull-requests
-->

## The Issue

- #6300
- #6246
- #5058

Composer options were not respected by `composer run-script`.

## How This PR Solves The Issue

Makes `ddev composer create` fully compatible with all possible composer options. I removed all the hard-coded options.
Adds composer options to `composer run-script`.
Adds `--no-interaction` to CakePHP quickstart.

---

`--preserve-flags` (which doesn't add `--no-plugins` and `--no-scripts`) didn't work as expected, I decided to remove it.

This is not a breaking change, because `--preserve-flags` didn't work before (it was always `false`).

---

I moved `composer run-script post-create-project-cmd` outside the condition for `composer install`, to make it work the same way as in `composer create-project`.

---

If you passed `--no-scripts`, DDEV tried to run `composer run-script`, which was wrong behavior, fixed it.

---

If `--help` or `--version` flags were provided, composer showed it's help page or it's version, and DDEV tried to continue creating the project, which was wrong behavior, fixed it.

---

I refactored the test logic to make it more clear and added more tests. I removed the testing for the `psr/log` package and replaced it with custom-made fake packages.

## Manual Testing Instructions

1. The order of commands executed for `ddev composer create` must be the same as for `composer create-project`.

To check this behavior, you can use a test repository (it is used for `testdata` here) which contains all the necessary situations to check:

```
cd ~/workspace && mkdir composer-create && cd composer-create
ddev config --auto
mkdir -p .ddev/test-ddev-composer-create/{test-ddev-require,test-ddev-require-dev}

curl -o .ddev/test-ddev-composer-create/composer.json https://raw.githubusercontent.com/stasadev/ddev/20240612_stasadev_composer_create/cmd/ddev/cmd/testdata/TestComposerCreateCmd/.ddev/test-ddev-composer-create/composer.json
curl -o .ddev/test-ddev-composer-create/test-ddev-require/composer.json https://raw.githubusercontent.com/stasadev/ddev/20240612_stasadev_composer_create/cmd/ddev/cmd/testdata/TestComposerCreateCmd/.ddev/test-ddev-composer-create/test-ddev-require/composer.json
curl -o .ddev/test-ddev-composer-create/test-ddev-require-dev/composer.json https://raw.githubusercontent.com/stasadev/ddev/20240612_stasadev_composer_create/cmd/ddev/cmd/testdata/TestComposerCreateCmd/.ddev/test-ddev-composer-create/test-ddev-require-dev/composer.json

export repo='{"type": "path", "url": ".ddev/test-ddev-composer-create", "options": {"symlink": false}}'

# see the result for:
composer create-project --repository $repo test/ddev-composer-create # add different flags here

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf

# see the result for:
ddev composer create --repository $repo test/ddev-composer-create # add different flags here

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf
````

`test/ddev-composer-create` package (`.ddev/test-ddev-composer-create/composer.json`) contains this:
- it requires `test/ddev-require` from `require` section:
  - `vendor/test/ddev-require` will be here, but only if there is no `--no-install` flag
- it requires `test/ddev-require-dev` from `require-dev` section:
  - `vendor/test/ddev-require-dev` will be here, but only if there is no `--no-dev`/`--no-install` flag
- it runs `post-root-package-install` script:
  - this script should run before `composer install`
  - it creates `created-by-post-root-package-install` file, but only if there is no `--no-scripts` flag
- it runs `post-create-project-cmd` script:
  - this script should run after `composer install`
  - it creates `created-by-post-create-project-cmd` file, but only if there is no `--no-scripts` flag

And use different flag combinations to see how it works for `composer create-project` and `ddev composer create`. It is important to check if the commands run in the same order.

Available flags for `composer create-project` https://getcomposer.org/doc/03-cli.md#create-project

Ignore the `test-ddev-require` and `test-ddev-require-dev` directories created in the project root (they are expected to be created every time as they are part of the composer package).

Every `ddev composer create` will have this output:
```
Executing composer command: [composer ...]
```
You can see what flags are used for each of these composer commands, and flags like `--no-interaction` will be also used for `composer run-script` and `composer install`.

For example:

This should not create `vendor` folder, and there should be no `created-by-*` files
```
# see the result for:
composer create-project --repository $repo test/ddev-composer-create --no-install --no-scripts

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf

# see the result for:
ddev composer create --repository $repo test/ddev-composer-create --no-install --no-scripts

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf
```

This should create `vendor/test/ddev-require` folder, (but not `vendor/test/ddev-require-dev`) and there should be no `created-by-*` files
```
# see the result for:
composer create-project --repository $repo test/ddev-composer-create --no-scripts --no-dev

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf

# see the result for:
ddev composer create --repository $repo test/ddev-composer-create --no-scripts --no-dev

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf
```

And so on.

---

2. Try CakePHP quickstart (without interaction), it should not ask for input:

```
mkdir my-cakephp-site && cd my-cakephp-site
ddev config --project-type=cakephp --docroot=webroot
ddev composer create --prefer-dist --no-interaction cakephp/app:~5.0
```

---

3. Try `Steps To Reproduce` example (it should not ask for input) from:

- #6246

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

rfay reviewed (changes requested):
Could you please add test coverage for this? Thanks! We have been fiddling with `ddev composer create` for years and years now, and we keep incrementally improving it, but we never seem to completely get it finished.

stasadev commented:
Yes, I will add some new test or modify an existing one.

stasadev commented on cmd/ddev/cmd/composer-create.go:
```diff
@@ -43,9 +44,22 @@ ddev composer create --preserve-flags --no-interaction psr/log
 			osargs = os.Args[3:]
-			osargs = nodeps.RemoveItemFromSlice(osargs, "--yes")
-			osargs = nodeps.RemoveItemFromSlice(osargs, "-y")
+			// I don't know why, but cmd.Flags().GetBool("preserve-flags") is always false, so we can check it here
+			preserveFlags = slices.Contains(osargs, "--preserve-flags")
```
This baffles me, I don't understand why Cobra reads `--preserve-flags` as `false` no matter how I try to pass this flag.

Anyway, at least `--preserve-flags` now works as described (it didn't before).

stasadev commented:
I added more conditions and checks to the test, now it covers most possible situations.

tyler36 commented:
Confirmed. This PR fixes #6300 .  Thank you for the quick turn around.

## Test

1. Confirm DDEV version

```shell
$ ddev -v
ddev version v1.23.1-62-g4f748c85a
```

1. Follow (updated) cakePHP quickstart. 

```shell
mkdir my-cakephp-site && cd my-cakephp-site
ddev config --project-type=cakephp --docroot=webroot
ddev composer create --prefer-dist --no-interaction cakephp/app:~5.0
ddev cake
ddev launch
```

> [!NOTE]  
This PR updates the quickstart to include `--no-interaction` . This makes no difference in DDEV `1.23.1` (see #6300 )

stasadev commented:
There is still some difference how composer scripts work in:

- `ddev composer create --no-install` runs only `post-root-package-install` and nothing else
- `composer create-project --no-install` runs `post-root-package-install` and `post-create-project-cmd`

I will investigate this further today and adjust the behavior.

stasadev commented:
`--preserve-flags` (which doesn't add `--no-plugins` and `--no-scripts`) didn't work as expected:

The order was:

- `composer create-project`
- `composer run-script post-root-package-install`
- `composer run-script post-create-project-cmd`
- `composer install`

The right order:

- `composer create-project`
- `composer run-script post-root-package-install`
- `composer install`
- `composer run-script post-create-project-cmd`

This is why I made `--preserve-flags` not add only `--no-plugins`

This is not a breaking change, because `--preserve-flags` didn't work before (it was always `false`).

---

I moved `composer run-script post-create-project-cmd` outside the condition for `composer install`, to make it work the same way as in `composer create-project`.

---

If you passed `--no-scripts`, DDEV tried to run `composer run-script`, which was wrong behavior, fixed it.

---

If `--help` or `--version` flags were provided, composer showed it's help page or it's version, and DDEV tried to continue creating the project, which was wrong behavior, fixed it.

---

I refactored the test logic to make it more clear and added more tests. I removed the testing for the `psr/log` package and replaced it with custom-made fake packages.

stasadev commented:
When I split combined short flags, it broke `-vvv` functionality.
It turned out that `-v -v -v` is not the same as `-vvv`, so I will see how to fix it.

stasadev commented:
Rebased.

I fixed `-vvv` and changed a bit the logic for flags.

I updated our documentation with a more detailed explanation of what steps we do with `ddev composer create`.

I researched the Composer documentation, learned more about what `--no-plugin`, `--no-scripts` and `--no-install` flags do. In result I decided to remove `--preserve-flags` completely, because if we do not add `--no-plugins`, `--no-scripts` and `--no-install` to `composer create-project`, it breaks the original logic, so I don't think we ever needed it.

`--preserve-flags` didn't work before this PR, so this removal is not a breaking change.

---

I'm pretty sure the most of the original logic here works right.

There are some conflicts with the `composer create-project` flags, such as `--ask` and `--working-dir`, using which results in error for `ddev composer create`, but I decided not to hard-code them, as the composer may add/remove some flags in the future, and I don't want to add unnecessary checks.

rfay commented on docs/content/users/usage/developer-tools.md:
```diff
@@ -40,7 +40,14 @@ For example:
 When using `ddev composer create` your project should be essentially empty or the command will refuse to run, to avoid loss of your files.
 
-By default `--no-plugins` and `--no-scripts` options are used while cloning the project and `composer install` is executed afterwards as long as `--no-install` is not provided. When using `--preserve-flags` plugins and scripts are allowed during the cloning step.
+Standard order of operations for `ddev composer create`:
```
I sure like this section!

```suggestion
`ddev composer create` does these things in this order:
```

rfay commented on docs/content/users/usage/developer-tools.md:
```diff
@@ -40,7 +40,14 @@ For example:
+Standard order of operations for `ddev composer create`:
+
+1. `composer create-project --no-plugins --no-scripts --no-install` clones a bare Composer package without any additional actions.
+2. `composer run-script post-root-package-install` runs if `--no-scripts` is not provided manually.
```
```suggestion
2. `composer run-script post-root-package-install` runs if `--no-scripts` is not given as a flag to `ddev composer create`.
```

rfay commented on docs/content/users/usage/developer-tools.md:
```diff
@@ -40,7 +40,14 @@ For example:
+
+1. `composer create-project --no-plugins --no-scripts --no-install` clones a bare Composer package without any additional actions.
+2. `composer run-script post-root-package-install` runs if `--no-scripts` is not provided manually.
+3. `composer install` runs if `--no-install` is not provided manually.
```
```suggestion
3. `composer install` runs if `--no-install` is not given as a flag to `ddev composer create`.
```

rfay commented on docs/content/users/usage/developer-tools.md:
```diff
@@ -40,7 +40,14 @@ For example:
+1. `composer create-project --no-plugins --no-scripts --no-install` clones a bare Composer package without any additional actions.
+2. `composer run-script post-root-package-install` runs if `--no-scripts` is not provided manually.
+3. `composer install` runs if `--no-install` is not provided manually.
+4. `composer run-script post-create-project-cmd` runs if `--no-scripts` is not provided manually.
```
```suggestion
4. `composer run-script post-create-project-cmd` runs if `--no-scripts` is not given as a flag to `ddev composer create`.
```

rfay reviewed (commented):
I just looked through the docs, not code, had minor comments.

rfay commented:
@rpkoller would you be willing to validate the quickstarts with this PR? I know it's loads of work, but your attention is so valuable.

rpkoller commented:
> @rpkoller would you be willing to validate the quickstarts with this PR? I know it's loads of work, but your attention is so valuable.

argh that slipped through again. was on my to do list then i completely forgot. will try to run through either tomorrow or on thursday hopefully.

rfay commented on cmd/ddev/cmd/composer-create.go:
```diff
@@ -27,25 +27,44 @@ var ComposerCreateCmd = &cobra.Command{
-			osargs = nodeps.RemoveItemFromSlice(osargs, "--preserve-flags")
+			var expandedOsargs []string
+			for _, osarg := range osargs {
+				// If this is a combined short option like "-test", split it into "-t -e -s -t"
```
Weird that we have to do this, when Cobra has these capabilities. I know there's *lots* of history in this code.

stasadev replied:
I don't think we are using Cobra here, because `osargs = os.Args[3:]` at the top
In any case, there is always a way to improve it in the future.

rfay commented on cmd/ddev/cmd/composer-create.go:
```diff
@@ -213,38 +280,9 @@ ddev composer create --preserve-flags --no-interaction psr/log
-						composerCmd = append(composerCmd, osarg)
-					}
+			for _, validCreateArg := range validCreateArgs {
+				if isValidComposerOption("install", validCreateArg) {
```
Thanks for splitting this out.

rfay commented on cmd/ddev/cmd/composer-create.go:
```diff
@@ -329,8 +379,6 @@ for basic project creation or 'ddev ssh' into the web container and execute
 }
 
 func init() {
-	ComposerCreateCmd.Flags().BoolP("yes", "y", false, "Yes - skip confirmation prompt")
```
It's always great to get evil things out of the init() functions

stasadev commented:
> we're in hell here and we're going to have to figure out how to get out of it

It sounds like a challenge and I like it. I haven't touched any code related to the mutagen, a few restarts here seems redundant. But still, I think I'll come back to it some time.

> Note that a --no-interaction install of Craft CMS does not result in a working site, because it still needs to do a ddev craft setup to work. (#6246)

I don't think we can update this quickstart with @mandrasch's technique right now, but after some time when this code is mature, we can update it for `--no-interaction`.

mandrasch commented:
> The original idea was just to allow composer create-project without doing it in a subdirectory. I'm not sure why we didn't end up doing all the work in a temporary directory and then copy it into the project, and I'm not sure why we can't do that now.

Just as sidenote: I wondered about this when I first tried to understand why there is a special command for composer create in ddev. But I haven't had time to find out why (seems like there is git involved, so directory really must be empty?  https://github.com/composer/composer/issues/7455#issuecomment-536991910) 🤓 Thanks for all the effort which went into making this work! 🙏

mandrasch commented:
Regarding CraftCMS, it's a bit confusing - the official docs state this (https://craftcms.com/docs/5.x/install.html#quick-start):

```bash
mkdir my-craft-project
cd my-craft-project/
ddev config --project-type=craftcms --docroot=web --php-version=8.2
ddev composer create -y --no-scripts "craftcms/craft"
ddev craft install
```

Using `--no-scripts` means, the post-create hook with `craft setup` is never used anyway. And at least I always used `ddev craft install` afterwards as well. There is no need for `--no-interaction` so far when following the official docs approach. Hope this does not lead to more confusion!

rpkoller commented on docs/content/users/quickstart.md:
```diff
@@ -54,7 +54,7 @@ Please note that you will need to change the PHP version to 7.4 to be able to wo
     ddev config --project-type=cakephp --docroot=webroot
-    ddev composer create --prefer-dist cakephp/app:~5.0
+    ddev composer create --prefer-dist --no-interaction cakephp/app:~5.0
     ddev cake
```
```suggestion
```
i think the `ddev cake` line can be dropped since it is missing an actual command doing something. at the moment it results in:  ```$> ddev cake
No command provided. Choose one of the available commands.```

the cake php docs also dont use the cake command setting up a new project: https://book.cakephp.org/5/en/installation.html#installation-using-ddev

stasadev replied:
Yes, it makes sense to remove that line, thanks.

stasadev commented:
@rfay and I discussed our current implementation for `ddev composer create`, and decided to try using `composer create-project` as is.

I did that and it generally worked, but when I started testing it with our quickstarts, things went wrong.

Our current logic:
- run `composer create-project --no-plugins --no-scripts --no-install`, which creates the project skeleton including important CMS settings, like `.env.example`.
- run `composer run-script post-root-package-install`
- restart the project, which creates `.env` from `.env.example`, *a really important step* for some frameworks like Laravel and CraftCMS
- run `composer install`
- run `composer run-script post-create-project-cmd`, which *heavily depends on CMS settings* in `.env` for some frameworks like Laravel and CraftCMS

For example:

Laravel quickstart creates `.env` with settings for sqlite, and runs database migrations. Without manual `composer run-script post-create-project-cmd` we don't run it for our mariadb.

CraftCMS asks some interactive questions and without DDEV override it suggested me this:
```
Database server name or IP address: [127.0.0.1]
```
instead of
```
Database server name or IP address: [db]
```

So we are tied to the current implementation.

rfay commented:
Thanks for studying so carefully!

stasadev commented:
> DDEV does the magic for .env in ddev restart (I also wondered why ddev's composer create needs a ddev restart)

And it also restarts Mutagen for macOS users.

---

> faster installations without a ddev restart

We could probably review what we do inside `ddev restart` for the CMS settings and do it without restarting.
The problem is that when we start the project, it checks many different settings, so it's just easier to restart it.

---

> install cms with CMSes regular CLI command, evaluates `.env` and so prefill should be correct ddev value
> `ddev craft install/craft` or `typo3 install:setup`

For CraftCMS: it runs `php craft setup/welcome` inside `post-create-project-cmd` script (when you run in without `--no-scripts`)
For Laravel: it runs `php artisan migrate` inside `post-create-project-cmd` script
As so on.

If we move to the original behavior for `composer create-project`:

1. this will mean that some actions will have to be performed twice (as in `post-create-project-cmd`). Of course, we can use a flag for `--no-scripts`, but this will change the initial behavior for `composer create-project`.
2. we will need to add all these framework-specific commands such as `ddev craft install/craft`, `ddev artisan migrate` and so on to our quickstarts again, which will increase the maintenance burden on DDEV. Today the framework has this set of commands, tomorrow they will add something else, so I don't like this idea. Notice how we cleaned up our quickstarts so that they will work all the time, even if the framework changes something inside the `post-create-project-cmd`. And we achieve this only because we intercept the `composer create-project` halfway before it executes the `post-create-project-cmd`.

mandrasch commented:
Great additions, thanks for insights @stasadev 👍👍 

After reading your post, I noticed the (third) big question is: Does DDEV want devs to follow DDEV quickstart or the official docs?

Because https://laravel.com/docs/11.x/installation for instance also says "run migrate if you need db". 

My personal experience is that DDEV automagic also might hide some important commands which are needed for deployment later on (but this argument is only true for inexperienced beginners + I wasn't aware of all the post create hooks in Composer since there were not executed in the past in DDEV)

mandrasch commented:
PS: I also wonder what framework/CMS creators would prefer - no scripts or with scripts? 🤔 

PPS: When I think about a DDEV UI in a far future (maybe some day 😁), this would also be a question - guess no-interaction would be needed here. 

(And thanks for the exchange, super interesting to learn more about how this works👍)

stasadev commented:
> Does DDEV want devs to follow DDEV quickstart or the official docs?

That's a good question. Usually, the `composer create-project` configures everything for you, but sometimes it doesn't 🙂 (for example, Drupal requires more actions). I think we need to have a certain balance, if the framework provides everything in `composer create-project` then we rely on it completely, otherwise we add more steps to our quickstarts.

> Because https://laravel.com/docs/11.x/installation for instance also says "run migrate if you need db".

Yes, it's in the Laravel documentation and also in the `post-create-project-cmd` script. They added it to the script quite recently, after they switched to SQLite in Laravel 11.

rfay commented:
> Does DDEV want devs to follow DDEV quickstart or the official docs?

We want devs to be successful in their own world, and be comfortable there. The quickstarts are normally adaptations of the official docs. But where the official docs don't reflect a docker-based environment (or reflect an opinionated one of their own) we have to adapt (and usually simplify).

Sometimes we do too much for the user. Drupal users can get so complacent about what DDEV does with settings that they forget how the whole settings system works and then they don't know how to debug it, which a few years ago would have been natural and easy for every user.
